ssas/service/blacklist/blacklist.go

```diff
@@ -0,0 +1,325 @@
+		ssas.CacheSyncFailure(ssas.Event{Op: "BlacklistLoadGroupsFromDatabase", TrackingID: trackingID, Help: err.Error()})
+	} else {
+		ssas.OperationSucceeded(ssas.Event{Op: "InitGroupBlacklist", TrackingID: trackingID})
+	}
+
+	bl.tcTicker, bl.tcDone = bl.startCacheRefreshTicker(tokenCacheRefresh, bl.loadTokensFromDatabase)
+	bl.gcTicker, bl.gcDone = bl.startCacheRefreshTicker(groupCacheRefresh, bl.loadGroupsFromDatabase)
+
+	return &bl
+}
+
+type blacklist struct {
+	sync.RWMutex
+
+	tc         *cache.Cache
+	tcLifetime time.Duration
+	tcTicker   *time.Ticker
+	tcDone     chan struct{}
+
+	gc         *cache.Cache
+	gcLifetime time.Duration
+	gcTicker   *time.Ticker
+	gcDone     chan struct{}
+
+	ID string
+}
+
+type groupEntry struct {
+	field      GroupField
+	expression string
+}
+
+//	BlacklistToken invalidates the specified tokenID
+func (b *blacklist) BlacklistToken(tokenID string) error {
+	entryDate := time.Now()
+	expirationDate := entryDate.Add(b.tcLifetime)
+	if _, err := createTokenEntry(tokenID, entryDate, expirationDate); err != nil {
+		return fmt.Errorf(fmt.Sprintf("unable to blacklist token id %s: %s", tokenID, err.Error()))
+	}
+
+	// Add to cache only after token is blacklisted in database
+	ssas.TokenBlacklisted(ssas.Event{Op: "TokenBlacklist", TrackingID: tokenID, TokenID: tokenID})
+	b.tc.Set(tokenID, entryDate.Unix(), b.tcLifetime)
+
+	return nil
+}
+
+//	IsTokenBlacklisted tests whether this tokenID is in the blacklist cache.
+//	- Tokens should expire before blacklist entries, so a tokenID for a recently expired token may return "true."
+//	- This queries the cache only, so if a tokenID has been blacklisted on a different instance, it will return "false"
+//		until the cached blacklist is refreshed from the database.
+func (b *blacklist) IsTokenBlacklisted(tokenID string) bool {
+	b.RLock()
+	defer b.RUnlock()
+
+	bEvent := ssas.Event{Op: "TokenVerification", TrackingID: b.ID, TokenID: tokenID}
+	if _, found := b.tc.Get(tokenID); found {
+		ssas.BlacklistedTokenPresented(bEvent)
+		return true
+	}
+	return false
+}
+
+func (b *blacklist) BlacklistGroup(field GroupField, value string) error {
+	entryDate := time.Now()
+	expirationDate := entryDate.Add(b.gcLifetime)
+	entry, err := createGroupEntry(value, entryDate, expirationDate, field)
+	if err != nil {
+		return fmt.Errorf(fmt.Sprintf("unable to create group blacklist entry value %s: %s", value, err.Error()))
+	}
+
+	// Add to cache only after token is blacklisted in database
+	ssas.GroupBlacklisted(ssas.Event{Op: "GroupBlacklist", TrackingID: strconv.FormatUint(uint64(entry.ID), 10)})
+
+	b.gc.Set(uuid.New(), groupEntry{field, value}, b.gcLifetime)
+
+	return nil
+}
+
+func (b *blacklist) IsGroupBlacklisted(group ssas.Group) bool {
+	var cacheItems map[string]cache.Item
+
+	b.RLock()
+	cacheItems = b.gc.Items()
+	b.RUnlock()
+
+	var xDataObj map[string]interface{}
+	if err := json.Unmarshal([]byte(group.XData), &xDataObj); err != nil {
```

Comment:
Historically we've tried to keep SSAS unaware of the contents of `group.XData`.  Yes, we are expecting it to be in JSON format, but we're accepting text and so far the only manipulation we're doing is `strconv.Unquote()`.  For the purposes of the blacklist, what if we:

- Looked at XData as a string, not JSON
- Ignored the key value we're expecting (`cmsids`)
- Just searched to see whether the group ID (OrgID) in the group blacklist was present in the string

The benefit would be that SSAS continues knowing as little as possible about this data packet being passed along.

Reply:
> Historically we've tried to keep SSAS unaware of the contents of `group.XData`. Yes, we are expecting it to be in JSON format

Definitely agreed that we should continue keeping the XData field opaque to SSAS. The only extent that we're inspecting the field is to verify that the field can be deserialized into a JSON object. The actual logic checks can be found in the `checkXData` function. The implementation does not make any assumptions on what is contained in the JSON object. Instead walks through the tree and attempts to find a matching entry in our blacklist tree.

> * Looked at XData as a string, not JSON
> * Ignored the key value we're expecting (cmsids)

Might not be fully understanding, but I'm not sure how we would know what key values are present and should be ignored. If we put in a list of those keys (like cmsids), I think we're actually adding more knowledge of what is going into the xdata.

> Just searched to see whether the group ID (OrgID) in the group blacklist was present in the string

Definitely! Our schema designates a the Group field to check against. For the near short term, we want to support xdata checks. We should look at expanding it to other fields if we there's a need.

Reply:
Yeah, that wasn't clear.  Here's another way to say it: SSAS doesn't care what's in XData.  It shouldn't care what format it's in either.  A mostly-data-blind blacklist can be implemented by blacklisting a string that _shall not_ be present in XDATA--in this case, the string would be an org id.

Reply:
The format of XData is a contract between the other parties using SSAS.